### PR TITLE
feat: separate console identity from API credentials

### PR DESCRIPTION
**Pull request by delacjus**
## Description

Config entries previously tied console identity directly to `CONF_HOST` (the IP address or hostname used during setup). If a console's IP changed (due to DHCP renewal, router reassignment, or network subnet migration), the integration could lose track of the existing config entry or create duplicate entries.

This PR decouples console identity from API network credentials and provides automatic backfilling for existing entries:

1. **Hardware-backed Console Identity:**
   - Establishes persistent console identity decoupled from `CONF_HOST`.
   - Adds seamless backfill during `async_setup_entry` for existing config entries that predate console identity separation.

2. **Hardware payload matching (`model` and `type` support):**
   - Verified on live UniFi hardware: real console payloads populate `model` while leaving `type` unset and omitting `is_gateway`. Initial detection code that matched solely on `type` fell through to the host IP.
   - Updated detection logic to check both `model` and `type` against a shared token set in `const.py`, ensuring reliable hardware identity binding even when `type` is missing.
   - Fallback logic prefers valid `site_id` over bare host IP as a last resort.

3. **Code quality & cleanup:**
   - Formatted `config_flow.py` to resolve linter line length issues.
   - Cleaned up shadowed duplicate imports.

## Type of Change

- [ ] 🐛 **Bug fix** (non-breaking change fixing an issue)
- [x] ✨ **New feature** (non-breaking change adding functionality)
- [ ] 💥 **Breaking change** (fix or feature causing existing functionality/automations to break)
- [ ] ♻️ **Refactoring** (code organization, no functional change)
- [ ] 📝 **Documentation** (documentation updates only)
- [x] 🧪 **Tests** (adding or improving test coverage)
- [ ] 🔧 **Maintenance / Chore** (dependency updates, tool config, CI)

## Architecture & Home Assistant Quality Scale Checklist

- [x] **UniFi Developer Portal**: Checked [developer.ui.com](https://developer.ui.com/) for the latest official UniFi API documentation, endpoints, and schema capabilities to ensure alignment with what UniFi offers to developers.
- [x] **Data Flow**: Entities read solely from `coordinator.data` (no direct HTTP/network calls in entity properties).
- [x] **API Encapsulation**: Actions and endpoints route through coordinator methods or vendored API clients (`custom_components/unifi_insights/api/`).
- [x] **Base Entities**: Inherits from `UnifiInsightsEntity` or `UnifiProtectEntity` with `_attr_has_entity_name = True`.
- [x] **Parallel Updates**: `PARALLEL_UPDATES = 0` for coordinator/read-only platforms, `1` for action platforms.
- [x] **Error Handling**: API errors wrapped into user-friendly `HomeAssistantError` or `ServiceValidationError`.
- [x] **Unique IDs & Device Info**: Unique IDs are deterministic and devices properly attached to `DeviceInfo`.
- [x] **Translations**: New UI strings added to `strings.json` with appropriate translation keys.

## Validation Checklist

- [x] `script/lint` (or `ruff check .` and `ruff format .`) passed with 0 errors.
- [x] `mypy custom_components/unifi_insights` passed with 0 errors.
- [x] `bandit -r custom_components/unifi_insights` passed with 0 vulnerabilities.
- [x] `pytest` passed with ≥ 90% branch coverage (`pytest --cov=custom_components/unifi_insights`) and Codecov coverage targets met.
- [ ] `CHANGELOG.md` updated under `[Unreleased]` with clear user-facing descriptions.
- [x] Tested on a live local Home Assistant instance (`./script/develop`) without errors in logs.

## Breaking Changes (if applicable)

- [x] No breaking changes

## Additional Context

- **Live Hardware Tested:** Verified on live UniFi hardware (`deploy/pr-d-rev2`) with exact device payloads captured to guarantee that `model`-based detection accurately resolves real consoles.
- **Test Suite:** Added comprehensive test suite in `tests/test_console_identity.py` covering config flows, migrations, backfills, and device identification.

## Summary by CodeRabbit

* **New Features**
  * Console identities are now detected automatically from UniFi network and Protect devices.
  * Entries use console-specific names and identifiers, supporting multiple consoles under one cloud account.
  * Local and remote setup flows provide improved console selection and account validation.

* **Bug Fixes**
  * Existing entries are migrated while preserving stable identifiers, titles, options, and registry links.
  * Improved fallback handling when console details are incomplete or unavailable.

### CODE REVIEW

**Comment by delacjus** [comment]
@coderabbitai review

File: custom_components/unifi_insights/__init__.py [hunk 1]
```diff
@@ -224,6 +228,30 @@ def budget_left(budget: str) -> bool:
...
+    """Return a stable site id to identify a console that exposes no gateway."""
+    data = getattr(config_coordinator, "data", None)
+    if not isinstance(data, dict):
+        return None
+    sites = data.get("sites")
+    if not isinstance(sites, dict):
+        return None
+    for site_id in sites:
+        if isinstance(site_id, str) and site_id and site_id != "default":
+            return site_id
```
[the remark this answers is not in the document]
Comment (delacjus) [reply custom_components/unifi_insights/__init__.py:251]: Partially accepted. The conclusion is right — one console can end up with two entries — but the mechanism you identified isn't the likely route, so I fixed a different one in `774af5f`.

**On the reported mechanism.** The control flow is as described: `config_flow.py:239-244` falls back to `host.lower()` when `sites[0].id == "default"`, `config_flow.py:417-418` persists `CONF_CONSOLE_ID` only when a MAC was found, and `_first_site_id` skips `"default"` and picks the first remaining site. It is reachable — `api/network/models/site.py:42-44` synthesises `id` from `internalReference`, and the regression test for #80 pins the live payload `{"internalReference": "default", "name": "Default"}` → `Site.id == "default"`. But it additionally needs a second, non-default site, i.e. a multi-site controller. Narrow.

**The route that doesn't need any of that.** The two console detectors disagree regardless of site ids: `_is_console_device` honours `is_gateway` (`__init__.py:233`), and the flow's matcher never looked at it (0 occurrences in `config_flow.py`). A gateway whose model isn't in `CONSOLE_DEVICE_TOKENS` is therefore invisible to the flow while setup happily adopts its MAC — the flow derives a host-based `unique_id`, setup rewrites it to the MAC, and a re-add no longer matches in `_abort_if_unique_id_configured()`. Same duplicate-entry outcome, no multi-site controller required. Worth noting the PR's own test at `tests/test_console_identity.py:628` already asserts the `is_gateway` case matters — on the setup side only.

So the flow's matcher now checks `is_gateway`, mirroring `_is_console_device`. It's `is True` rather than truthy because `getattr` on a bare `MagicMock` returns a truthy mock, which would have made every device match in tests — the file already guards this way with its `type(x) is str` checks.

**On the proposed shared site-id helper:** I didn't do it. A helper that only unifies the *site-id* rule fixes the rare multi-site case and leaves the `is_gateway` divergence open. The real convergence is a shared console-identity predicate plus multi-site device scanning in the flow, and the latter adds N API calls to a config flow — a behaviour change, not a refactor. Tracking separately rather than growing this PR.

File: custom_components/unifi_insights/__init__.py [hunk 2]
```diff
@@ -524,3 +622,50 @@ async def async_reload_entry(
...
+
+    if config_entry.version == 1:
+        new_data = dict(config_entry.data)
+        new_unique_id = config_entry.unique_id
+
+        # Remote migration: ensure unique_id is the console_id
+        if (
+            new_data.get(CONF_CONNECTION_TYPE) == CONNECTION_TYPE_REMOTE
+            or new_data.get(CONF_CONNECTION_TYPE) == CONNECTION_TYPE_LOCAL
+        ):
```
[the remark this answers is not in the document]
Comment (delacjus) [reply custom_components/unifi_insights/__init__.py:648]: Not accepted — the control flow is right, but nothing in this repo's history can produce the state.

An entry needs `CONF_CONSOLE_ID` *without* `CONF_CONNECTION_TYPE` for the condition to skip it. Checking every writer:

- The original `data=user_input` form (`2e123ab:43`) predates `CONF_CONSOLE_ID` entirely — it first appears in `7148af3`. Legacy entries have **neither** key, so they aren't the case you describe.
- Every `async_create_entry` that writes `CONF_CONSOLE_ID` writes `CONF_CONNECTION_TYPE` in the same dict (`7148af3:171-172`, `c6de47d:313-314`, `2dde746:365-366`, `38b91db:365-366`, and `config_flow.py:543-545` at head).
- The one merge-style writer that could add `CONF_CONSOLE_ID` to `{**entry.data}` without the type key (`config_flow.py:645-649`) sits behind `entry.data.get(CONF_CONNECTION_TYPE, CONNECTION_TYPE_LOCAL)` (`config_flow.py:596-598`), so a key-less entry takes the local branch, which merges only `CONF_API_KEY`. Reconfigure has the same guard at `:694`.

Even for a hypothetical such entry the outcome is benign: `__init__.py:657-663` bumps `minor_version` unconditionally so the migration never re-runs, and `__init__.py:439-467` backfills on the next setup because a missing type already defaults to local — which is the same defaulting the suggested diff proposes, just applied a step later.

Separately, the condition as written is already vacuous: both arms of the `or` lead to identical behaviour, so the honest cleanup is to delete the branch rather than add a default to it. Not doing that here either, since it's unrelated to this PR's scope.

File: custom_components/unifi_insights/config_flow.py [hunk 3]
```diff
@@ -586,23 +696,41 @@ async def async_step_reconfigure(
...
                         return self.async_update_reload_and_abort(
                             entry,
-                            data={
-                                CONF_CONNECTION_TYPE: CONNECTION_TYPE_LOCAL,
-                                CONF_HOST: user_input[CONF_HOST],
-                                CONF_API_KEY: user_input[CONF_API_KEY],
-                                CONF_VERIFY_SSL: user_input.get(CONF_VERIFY_SSL, False),
-                            },
+                            data=new_data,
+                            unique_id=new_id or entry.unique_id,
```
[the remark this answers is not in the document]
Comment (delacjus) [reply custom_components/unifi_insights/config_flow.py]: Accepted, and it's worse than reported. Fixed in `774af5f`.

Reproduced: with `CONF_CONSOLE_ID = "aa:bb:cc:dd:ee:ff"` stored, making `devices.get_all` raise `TimeoutError` during reconfigure yields `reconfigure_successful` (no `account_mismatch`), and leaves `entry.data[CONF_CONSOLE_ID] == "site-a"` with `entry.unique_id == "site-a"`. The trigger is cheap: `config_flow.py:215-234` wraps the whole device inspection in `except Exception`, so any timeout, 5xx or rate-limit leaves no `"mac"`, and `:239-244` then fills `"id"` from the site or host. Reconfigure is exactly when the network is being changed, so transient failures are at their most likely.

Two things that make it worse than "the wrong value gets stored":

1. **It's permanent.** On the next setup, `console_id` is truthy and `!= api_key`, so the backfill at `__init__.py:439` is skipped. The real MAC is sitting in the device coordinator data and is never adopted.
2. **It compounds.** Once the id is no longer MAC-shaped, the `":"`-on-both guard can never fire again, so a later reconfigure pointed at a genuinely *different* console is accepted silently and `CONF_CONSOLE_ID` becomes that console's MAC. The `account_mismatch` protection is defeated from then on.

The fix is the one you suggested — only a discovered MAC may replace the stored identity:

```python
current_id = entry.data.get(CONF_CONSOLE_ID) or entry.unique_id
new_id = console_info.get("mac") or current_id or console_info.get("id")
```

The `":"`-on-both guard is unchanged, so a different discovered MAC still aborts. Added a regression test for the transient-`devices.get_all` case; both new tests fail against the previous code and pass against the fix.

Your "`new_id` is empty" sub-claim was also correct. `test_local_reconfigure_with_blank_host_yields_no_console_id` asserted the stored id gets dropped, which encoded the bug as intended behaviour — it's renamed to `..._preserves_console_id` and now asserts the opposite. `new_id` can still be falsy for a legacy entry with no stored id and no unique_id, so there's a new test covering that the key is omitted rather than written empty.

File: tests/test_console_identity.py [hunk 4]
```diff
@@ -0,0 +1,1756 @@
...
+            new_callable=AsyncMock,
+        ),
+        patch(
+            "homeassistant.config_entries.ConfigEntries.async_forward_entry_setups",
+            new_callable=AsyncMock,
+        ),
+    ):
+        mock_probe_net.return_value = ProbeResult(
+            ProbeStatus.AVAILABLE, [MagicMock(id="default")]
+        )
```
[the remark this answers is not in the document]
Comment (delacjus) [reply tests/test_console_identity.py:1037]: Accepted and fixed in `19ef68b`. `ProbeResult` is `(status, error=None, sites=[])`, so the list was landing in `error` and `network_probe.sites` stayed empty — those tests logged `found 0 sites` and never exercised the site-dependent behaviour their names describe.

All seven call sites now pass `sites=` by keyword (751, 823, 914, 1259, 1337, 1410, 1476). Worth recording where they came from, since it splits: 751, 823 and 914 predate this branch, and the other four were added with the coverage tests, which copied the existing pattern rather than checking the signature.

The suite stays green with `sites` actually populated, so the assertions were sound — they just weren't reaching the code they claimed to. Patch coverage is unchanged at 100%.

File: tests/test_console_identity.py [hunk 5]
```diff
@@ -0,0 +1,1756 @@
...
+
+
+async def test_console_identity_final_branch_coverage(
+    hass: HomeAssistant,
+) -> None:
+    """Cover remaining edge cases in config flow and protect coordinator."""
+    from custom_components.unifi_insights.coordinators.protect import (
+        UnifiProtectCoordinator,
+    )
+    from custom_components.unifi_insights.api import UniFiAuthenticationError
```
[the remark this answers is not in the document]
Comment (delacjus) [reply tests/test_console_identity.py]: Fixed in `19ef68b`.

One correction for the record: this doesn't fail CI. Ruff isn't wired into any workflow on this repo — the `validate` job is hassfest, and the only other jobs are the test/coverage run and HACS. `script/lint` is local-only, which is why an `F401` could sit on a green PR.

While in the file I also hoisted the `UniFiResponseError` import used by a test I added, so this change doesn't introduce a new `PLC0415`. The file's ruff output is now byte-identical to its state before these commits (30 findings, 21 `PLC0415`), so nothing new was added. The five pre-existing function-level imports of that same symbol are untouched, as cleaning them up would be unrelated churn in this PR.

File: custom_components/unifi_insights/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (delacjus) [reply custom_components/unifi_insights/__init__.py:251]: Tracked in #147 — it's a new issue, there was no existing one.

It covers both halves of the remaining divergence rather than just the site-id rule: the scan-scope gap (the flow inspects `sites[0]` only, while setup scans every site plus the Protect NVR) and the `"default"` handling you originally reported, including the `internalReference` → `"default"` payload from #80 that makes it reachable.

The proposed fix there is a shared `_resolve_console_identity(devices)` helper called by both paths, so the predicate and the id precedence are defined once instead of being kept in sync by hand — which is what let `is_gateway` drift in the first place.

Flagged in the issue, since it's the part worth deciding before anyone implements it: widening the flow's scan adds one `devices.get_all` call per site to the config flow. That's a real latency cost during setup, not a free refactor, which is the main reason it didn't ride along in this PR.

[the remark this answers is not in the document]
Comment (delacjus) [reply custom_components/unifi_insights/__init__.py:251]: Implemented in #148 (draft, stacked on this PR).

On the latency trade-off you flagged: it's avoided rather than accepted. The flow scans sites in API order and stops at the first console it finds, so the ordinary single-site console still costs exactly one `devices.get_all` call. Only a controller whose gateway sits outside the first site issues more — which is precisely the case that was broken. There's a test asserting the single-call behaviour so the scan can't quietly become eager later.

The shared module also turned out to need one thing worth naming: the two callers hold different shapes (setup reads coordinator dicts, the flow holds API models), so devices are projected through a `device_fields()` helper that copies a value only when it has the expected type. An unguarded `getattr` would read a test double's auto-created attributes and make every device look like a gateway — the same trap that made `is_gateway` need `is True` here.
